File: src/browser_interface/browser/probe.py
"""Optional same-browser diagnostic probe."""

import copy
from typing import Any

from shared import upstream_value

PROBE_SCRIPT = """async ({url, requestHeaders, requestPayload}) => {
  try {
    const response = await fetch(url, {
      method: "POST",
      credentials: "include",
      headers: requestHeaders,
      body: JSON.stringify(requestPayload),
    });
    return {status: response.status, body: await response.text()};
  } catch (error) {
    return {networkError: error instanceof Error ? error.message : String(error)};
  }
}"""

ALLOWED_HEADERS = {
    "accept", "accept-language", "authorization", "content-type",
    "x-browser-channel", "x-browser-copyright", "x-browser-validation",
    "x-browser-year", "x-client-data", "x-aistudio-visit-id",
    "x-goog-api-key", "x-goog-authuser",
    "x-goog-visitor-id", "x-user-agent",
}
ALLOWED_HEADERS.add(upstream_value("makersuite", "logging_context_header").lower())


async def probe_generate(
    page: Any,
    latest_rpc_headers: dict[str, Any],
    runtime: dict[str, Any],
    generate_request: dict[str, Any],
    token: str,
) -> dict[str, Any]:
    payload = generate_request.get("payload")
    url = generate_request.get("url")
    if not isinstance(payload, list) or not isinstance(url, str):
        raise ValueError("GenerateContent request context is invalid for probe")

    headers = _allowed(generate_request.get("headers", {}))
    headers.update(_allowed(latest_rpc_headers))
    headers["x-goog-api-key"] = runtime["apiKey"]
    headers["x-aistudio-visit-id"] = runtime["visitId"]
    headers["x-goog-authuser"] = str(runtime.get("authUser", "0"))

    request_payload = copy.deepcopy(payload)
    request_payload[4] = token
    return await page.evaluate(
        PROBE_SCRIPT,
        {"url": url, "requestHeaders": headers, "requestPayload": request_payload},
    )


def _allowed(headers: dict[str, Any]) -> dict[str, str]:
    return {
        str(name).lower(): str(value)
        for name, value in headers.items()
        if str(name).lower() in ALLOWED_HEADERS
    }

File: go/browserinterface/probe.go
package browserinterface

import (
	"encoding/json"
	"fmt"
	"strings"

	cdpRuntime "github.com/chromedp/cdproto/runtime"
	"github.com/chromedp/chromedp"
)

type probeResult struct {
	Status       int    `json:"status"`
	Body         string `json:"body"`
	NetworkError string `json:"networkError"`
	Location     string `json:"location"`
}

func (s *ChromeSession) Probe(generateRequest map[string]any, token string) (probeResult, error) {
	payload, ok := generateRequest["payload"].([]any)
	url, urlOK := generateRequest["url"].(string)
	if !ok || !urlOK {
		return probeResult{}, fmt.Errorf("GenerateContent request context is invalid for probe")
	}
	encoded, _ := json.Marshal(payload)
	var requestPayload []any
	if err := json.Unmarshal(encoded, &requestPayload); err != nil || len(requestPayload) < 5 {
		return probeResult{}, fmt.Errorf("GenerateContent payload is invalid for probe")
	}
	requestPayload[4] = token

	headers := probeHeaders(stringMap(generateRequest["headers"]))
	for name, value := range probeHeaders(s.observedHeaders()) {
		headers[name] = value
	}
	s.stateMu.RLock()
	runtime := s.runtime
	s.stateMu.RUnlock()
	headers["x-goog-api-key"] = runtime.APIKey
	headers["x-aistudio-visit-id"] = runtime.VisitID
	headers["x-goog-authuser"] = runtime.AuthUser

	var result probeResult
	input := map[string]any{"url": url, "headers": headers, "payload": requestPayload}
	encodedInput, _ := json.Marshal(input)
	expression := "(" + probeScript + ")(" + string(encodedInput) + ")"
	awaitPromise := func(params *cdpRuntime.EvaluateParams) *cdpRuntime.EvaluateParams {
		return params.WithAwaitPromise(true)
	}
	if err := chromedp.Run(s.ctx, chromedp.Evaluate(expression, &result, awaitPromise)); err != nil {
		return probeResult{}, err
	}
	return result, nil
}

func (s *ChromeSession) Fingerprint() string {
	s.stateMu.RLock()
	defer s.stateMu.RUnlock()
	return s.fingerprint
}

func stringMap(value any) map[string]string {
	result := map[string]string{}
	source, _ := value.(map[string]any)
	for name, raw := range source {
		result[name] = fmt.Sprint(raw)
	}
	if typed, ok := value.(map[string]string); ok {
		return typed
	}
	return result
}

func probeHeaders(source map[string]string) map[string]string {
	result := map[string]string{}
	for name, value := range source {
		lower := strings.ToLower(name)
		if probeHeaderNames[lower] || strings.HasPrefix(lower, "x-goog-ext-") {
			result[lower] = value
		}
	}
	return result
}

var probeHeaderNames = map[string]bool{
	"authorization": true, "content-type": true,
	"x-aistudio-visit-id": true, "x-goog-api-key": true,
	"x-goog-authuser": true, "x-goog-visitor-id": true,
	"x-user-agent": true,
}

const probeScript = `async function(input) {
  try {
    const response = await fetch(input.url, {
      method: "POST",
      credentials: "include",
      headers: input.headers,
      body: JSON.stringify(input.payload),
    });
    return {status: response.status, body: await response.text(), location: location.href};
  } catch (error) {
    return {
      networkError: error instanceof Error ? error.message : String(error),
      location: location.href,
    };
  }
}`
